larpandoracontent/LArHelpers/LArHierarchyHelper.h

```diff
@@ -0,0 +1,629 @@
+         *
+         *          If only folding back to leading shower particles, the hierarchy will largely mirror the standard MCParticle hierarchy,
+         *          but, when a shower particle is reached (for this purpose an electron or photon), this particle and all daughter
+         *          particles will be represented by a single leaf node.
+         *
+         *          If folding back to both primary and leading shower particles the hierarchy will again be rather flat, but in this case,
+         *          if a primary track-like particle (i.e. not an electron or photon) has a downstream shower particle then all downstream
+         *          particles above the shower-like particle will be folded into the primary node, but a new, daughter leaf node will be
+         *          created for the shower-like particle and all of its daughters, and a parent-child relationship will be formed between
+         *          the primary node and shower node.
+         *
+         *  @param  mcParticleList The list of MC particles with which to fill the hierarchy
+         *  @param  caloHitList The list of hits with which to fill the hierarchy
+         *  @param  foldToPrimaries Whether or not to fold daughter particles back to their primary particle
+         *  @param  foldToLeadingShowers Whether or not to fold daughter particles back to their leading shower particle
+         */
+        void FillHierarchy(const pandora::MCParticleList &mcParticleList, const pandora::CaloHitList &caloHitList,
+            const bool foldToPrimaries, const bool foldToLeadingShowers);
+
+        // Note - likely want to change this to retrieve the neutrino, the test beam particle, all primaries, all cosmics etc
+        /**
+         *  @brief  Retrieve the root nodes in this hierarchy
+         *
+         *  @return The root nodes in this hierarchy
+         */
+        const NodeVector& GetRootNodes() const { return m_rootNodes; }
+
+        /**
+         *  @brief  Retrieve a flat vector of the ndoes in the hierarchy
+         *
+         *  @param  nodeVector The output vector for the nodes in the hierarchy in breadth first order
+         */
+        void GetFlattenedNodes(NodeVector &nodeVector) const;
+
+        /**
+         *  @brief  Produce a string representation of the hierarchy
+         *
+         *  @return The string representation of the hierarchy
+         */
+        const std::string ToString() const;
+
+        /**
+         *  @brief  Check if this is a neutrino hierarchy.
+         *
+         *  @return Whether or not this is a neutrino hierarchy.
+         */
+        bool IsNeutrinoHierarchy() const { return m_pNeutrino != nullptr; }
+
+        /**
+         *  @brief  Check if this is a test beam hierarchy.
+         *
+         *  @return Whether or not this is a test beam hierarchy.
+         */
+        bool IsTestBeamHierarchy() const { return m_pNeutrino == nullptr; };
+
```

Comment:
This is really about distinguishing the neutrino case from the test beam case, rather than identifying separate neutrino, testbeam and cosmics hierarchies. The cosmics hierarchy can coexist with both of these cases.